// @gamestop/gs-cli-aws-auth/lib/credentials-helper.js
'use strict';

const os = require('os');
const fs = require('fs');
const ini = require('ini');

class CredentialsHelper {
  static writeAwsCredentialsToFile(credentials) {
    return new Promise((resolve, reject) => {
      const credsFileDir = `${os.homedir()}/.aws`;
      const credsFilePath = `${credsFileDir}/credentials`;

      if (!fs.existsSync(credsFileDir)) {
        console.log('WARNING: aws credentials directory was not found!!! Creating...');
        fs.mkdirSync(credsFileDir);
      }

      if (!fs.existsSync(credsFilePath)) {
        console.log('WARNING: aws credentials file was not found!!! Creating...');
        fs.closeSync(fs.openSync(credsFilePath, 'w'));
      }

      const credsFileObject = ini.decode(fs.readFileSync(credsFilePath, 'utf-8'));
      credsFileObject.default = {
        output: 'json',
        region: 'us-east-1',
        aws_access_key_id: credentials.AccessKeyId,
        aws_secret_access_key: credentials.SecretAccessKey,
        aws_session_token: credentials.SessionToken
      };
      fs.writeFile(credsFilePath, ini.stringify(credsFileObject), (er) => {
        if (er) reject(er);
        else resolve(true);
      });
    });
  }
}

module.exports = CredentialsHelper;
